File: benchmarks/vecmath/VecMath.java
// vecmath — small-object temporary stress test (mirrors vecmath.rl).
// Java's natural shape is an object per vector; each operation allocates a
// fresh Vec3. The JIT's escape analysis is expected to scalar-replace the
// short-lived temporaries — that capability is part of what this measures.
public class VecMath {
    static final class Vec3 {
        final double x, y, z;

        Vec3(double x, double y, double z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        Vec3 cross(Vec3 b) {
            return new Vec3(
                y * b.z - z * b.y,
                z * b.x - x * b.z,
                x * b.y - y * b.x);
        }

        Vec3 addMod(Vec3 b, double m) {
            return new Vec3(
                (((x + b.x) % m) + m) % m,
                (((y + b.y) % m) + m) % m,
                (((z + b.z) % m) + m) % m);
        }
    }

    static long lcgNext(long state) {
        return (state * 1664525L + 1013904223L) % 4294967296L;
    }

    public static void main(String[] args) {
        final long T = 8000000;
        final double M = 1048576.0;

        long state = 42;
        Vec3 acc = new Vec3(0.0, 0.0, 0.0);

        for (long t = 0; t < T; t++) {
            state = lcgNext(state);
            double a1 = (double) (state & 1023);
            state = lcgNext(state);
            double a2 = (double) (state & 1023);
            state = lcgNext(state);
            double a3 = (double) (state & 1023);
            Vec3 v1 = new Vec3(a1, a2, a3);

            state = lcgNext(state);
            double b1 = (double) (state & 1023);
            state = lcgNext(state);
            double b2 = (double) (state & 1023);
            state = lcgNext(state);
            double b3 = (double) (state & 1023);
            Vec3 v2 = new Vec3(b1, b2, b3);

            Vec3 c = v1.cross(v2);
            acc = acc.addMod(c, M);
        }

        long checksum = (long) acc.x * 3 + (long) acc.y * 5 + (long) acc.z * 7;
        System.out.println(checksum);
    }
}

File: benchmarks/vecmath/vecmath.rs
// vecmath — small-object temporary stress test (mirrors vecmath.rl).
// Rust's natural shape is a Copy struct: methods return fresh Vec3 VALUES
// the compiler keeps in registers.

#[derive(Clone, Copy)]
struct Vec3 {
    x: f64,
    y: f64,
    z: f64,
}

impl Vec3 {
    fn cross(self, b: Vec3) -> Vec3 {
        Vec3 {
            x: self.y * b.z - self.z * b.y,
            y: self.z * b.x - self.x * b.z,
            z: self.x * b.y - self.y * b.x,
        }
    }

    fn add_mod(self, b: Vec3, m: f64) -> Vec3 {
        Vec3 {
            x: (((self.x + b.x) % m) + m) % m,
            y: (((self.y + b.y) % m) + m) % m,
            z: (((self.z + b.z) % m) + m) % m,
        }
    }
}

fn lcg_next(state: i64) -> i64 {
    (state.wrapping_mul(1664525).wrapping_add(1013904223)) % 4294967296
}

fn main() {
    const T: i64 = 8000000;
    const M: f64 = 1048576.0;

    let mut state: i64 = 42;
    let mut acc = Vec3 { x: 0.0, y: 0.0, z: 0.0 };

    for _ in 0..T {
        state = lcg_next(state);
        let a1 = (state & 1023) as f64;
        state = lcg_next(state);
        let a2 = (state & 1023) as f64;
        state = lcg_next(state);
        let a3 = (state & 1023) as f64;
        let v1 = Vec3 { x: a1, y: a2, z: a3 };

        state = lcg_next(state);
        let b1 = (state & 1023) as f64;
        state = lcg_next(state);
        let b2 = (state & 1023) as f64;
        state = lcg_next(state);
        let b3 = (state & 1023) as f64;
        let v2 = Vec3 { x: b1, y: b2, z: b3 };

        let c = v1.cross(v2);
        acc = acc.add_mod(c, M);
    }

    let checksum = (acc.x as i64) * 3 + (acc.y as i64) * 5 + (acc.z as i64) * 7;
    println!("{}", checksum);
}
